Solutions/Marain.Workflow.Api.Services/Marain/Workflows/Api/Services/EngineService.cs

```diff
@@ -69,5 +78,105 @@ public async Task<OpenApiResult> StartWorkflowInstance(string tenantId, StartWor
             await workflowEngine.StartWorkflowInstanceAsync(body).ConfigureAwait(false);
             return this.CreatedResult();
         }
+
+        /// <summary>
+        /// Gets a workflow definition.
+        /// </summary>
+        /// <param name="tenantId">The tenant Id for the current request.</param>
+        /// <param name="workflowId">The workflow Id.</param>
+        /// <returns>The <see cref="Task" />.</returns>
+        [OperationId(GetWorkflowOperationId)]
+        public async Task<OpenApiResult> GetWorkflow(string tenantId, string workflowId)
+        {
+            ITenant tenant = await this.marainServicesTenancy.GetRequestingTenantAsync(tenantId).ConfigureAwait(false);
+            IWorkflowStore workflowStore = await this.workflowStoreFactory.GetWorkflowStoreForTenantAsync(tenant).ConfigureAwait(false);
+            Workflow workflow = await workflowStore.GetWorkflowAsync(workflowId).ConfigureAwait(false);
+            OpenApiResult result = this.OkResult(workflow, "application/json");
+
+            if (!string.IsNullOrEmpty(workflow.ETag))
+            {
+                result.Results.Add("ETag", workflow.ETag);
+            }
+
+            return result;
+        }
+
+        /// <summary>
+        /// Stores a workflow definition.
+        /// </summary>
+        /// <param name="tenantId">The tenant Id for the current request.</param>
+        /// <param name="body">The workflow to store.</param>
+        /// <returns>The <see cref="OpenApiResult"/> to return.</returns>
+        [OperationId(CreateWorkflowOperationId)]
+        public async Task<OpenApiResult> CreateWorkflow(string tenantId, Workflow body)
+        {
+            // New workflow definitions shouldn't contain an etag.
+            if (!string.IsNullOrEmpty(body.ETag))
+            {
+                return new OpenApiResult { StatusCode = 400 };
+            }
+
+            ITenant tenant = await this.marainServicesTenancy.GetRequestingTenantAsync(tenantId).ConfigureAwait(false);
+            IWorkflowStore workflowStore = await this.workflowStoreFactory.GetWorkflowStoreForTenantAsync(tenant).ConfigureAwait(false);
+
+            // The API operation is "create" but the store only supports Upsert. We need to see if there's a workflow
+            // with the specified ID already before we try and save it.
+            try
+            {
+                _ = await workflowStore.GetWorkflowAsync(body.Id).ConfigureAwait(false);
+
+                // If we're here, we have found a workflow with the same Id, so we will return a 409 Conflict
+                return this.ConflictResult();
+            }
+            catch (WorkflowNotFoundException)
+            {
+                // No-op; this means we can continue and store the new workflow definition.
+            }
+
+            await workflowStore.UpsertWorkflowAsync(body).ConfigureAwait(false);
```

Comment:
This isn't safe without an ETag on the Upsert.

Reply:
(And with an ETag on the upsert, you don't need the get check.)

Reply:
Fixing this involved some internal tweaks to the Cosmos workflow store - as we discussed offline, the Cosmos `Container.UpsertAsync` method doesn't honour `If-None-Match` options when storing new instances, so internally we are determining whether it's a "create" or "update" operation based on whether an etag is supplied and then making the correct call.